### PR TITLE
[ENH] Remove `X.shape` checking, single labeled data warnings from TSC and TSR end estimators.

### LINKED ISSUE #9341: [ENH] Single labeled dataset warning
Perhaps you can propagate this to all other existing `torch` classifiers as well, in a separate PR. (created issue as a reminder for myself)

_Originally posted by @RecreationalMath in https://github.com/sktime/sktime/pull/9298#discussion_r2779468790_

### PR DESCRIPTION
**Pull request by Faakhir30**
#### Reference Issues/PRs

- Closes #9341

~~#9451~~ (Seperated changes related to activation issue to #9622)

#### What does this implement/fix? Explain your changes.

From both `classification` and `regression` estimators implementing BaseClassifer or BaseDeepRegressorTorch:
- removes single labeled dataset warning.
- removes `X.shape` checking. Refs https://github.com/sktime/sktime/pull/9298#discussion_r2831914759
- ~~Fixes `BaseClassifierTorch` not respecting `activation` passed by classifier estimators.~~

#### Does your contribution introduce a new dependency? If yes, which one?

No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

File: sktime/classification/deep_learning/base/_base_torch.py
Comment (RecreationalMath) [inline sktime/classification/deep_learning/base/_base_torch.py:207]: There is a docstring discrepancy here; `underscore` is missing at the start of both of the private attributes' names.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 1]
```diff
@@ -324,7 +339,23 @@ def _validate_activation_criterion(self):
             self._validated_activation = None
         else:
             self._validated_criterion = self.criterion
-            self._validated_activation = self.activation
+
+            # if activation is not one of the recognized activation functions,
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/base/_base_torch.py]: The check being performed here is already present (a more complete version of this) in the `_instantiate_activation` in respective network classes.

Is there a specific reason for doing this here? Or perhaps you misunderstood the point in [this](https://github.com/sktime/sktime/pull/9298#discussion_r2821943875) comment.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 2]
```diff
@@ -601,7 +632,11 @@ def _predict_proba(self, X):
                 y_pred.append(self.network(**inputs).detach())
         y_pred = cat(y_pred, dim=0)
         # (batch_size, num_outputs)
-        y_pred = Fsoftmax(y_pred, dim=-1)
+
+        if self._validated_activation:
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/base/_base_torch.py]: I believe this suffices as the intended solution for the point raised [here](https://github.com/sktime/sktime/pull/9298#discussion_r2821943875). 

https://github.com/Faakhir30/sktime/blob/d214b31548ff5a0667a24c2a64a3f8262a6d730f/sktime/classification/deep_learning/base/_base_torch.py#L342-L358 can be reverted to the original.

File: sktime/classification/deep_learning/rnn/_rnn_torch.py [hunk 3]
```diff
@@ -185,12 +185,6 @@ def _build_network(self, X, y):
...
         """
-        if len(X.shape) != 3:
-            raise ValueError(
-                f"Expected 3D input X with shape (n_instances, n_dims, series_length), "
-                f"but got shape {X.shape}. Please ensure your input data is "
-                "properly formatted."
-            )
         # n_instances, n_dims, n_timesteps = X.shape
         self.num_classes = len(np.unique(y))
         _, self.input_size, _ = X.shape
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/rnn/_rnn_torch.py:203]: Could you please check and comment: whether control can ever reach here in this line without going through the `_fit` (and having performed the required check that we are deleting here)?

File: sktime/regression/deep_learning/base/_base_torch.py [hunk 4]
```diff
@@ -112,6 +112,13 @@ def __init__(
         self._all_callbacks = None
 
     def _fit(self, X, y):
+        if len(X.shape) != 3:
```
Comment (RecreationalMath) [inline sktime/regression/deep_learning/base/_base_torch.py]: Please check if X has been transposed by the time control reaches here; if yes, then the shape is mis-specified in the error message.

Comment (RecreationalMath) [reply sktime/regression/deep_learning/base/_base_torch.py]: Same would apply for the base classifier class.

**Review by RecreationalMath — CHANGES_REQUESTED** [review]

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 1 above)
Comment (Faakhir30) [reply sktime/classification/deep_learning/base/_base_torch.py]: @RecreationalMath Correct me if I'm wrong:
- Issue we had is at `_predict` inference we are always using `Softmax` hardcoded, and it is not respecting non-default `activation` passed by estimator that falls into `else` block.
- Reason I added a check to make sure its `callable` here is because of the fact that `_instantiate_activation` is not being called in the base layer. So in this `else` block, If its some exotic `activation` we need to have its callable instance at inference time. 

Perhaps a better way of doing this in this is keeping `else` block like you've suggested.

But then also:
-  add to else block `self._validated_activation_instance=self.netowrk._instantiate_activation(self.activation)` 
- change the `_predict_proba` activation calling to:

```python
        if self._validated_activation_instance:
           
            y_pred = self._validated_activation_instance(y_pred)
        else:
            y_pred = Fsoftmax(y_pred, dim=-1)
```

File: sktime/regression/deep_learning/base/_base_torch.py (hunk 4 above)
Comment (Faakhir30) [reply sktime/regression/deep_learning/base/_base_torch.py]: Did check this one ahead, shape is changed in `_fit` after this check, so I think error msg is fine.

```python
    def _fit(self, X, y):
        if len(X.shape) != 3:
            raise ValueError(
                f"Expected 3D input X with shape (n_instances, n_dims, series_length), "
                f"but got shape {X.shape}. Please ensure your input data is "
                "properly formatted."
            )

        if len(np.unique(y)) == 1:
            warnings.warn(
                f"The provided data passed to {self.__class__.__name__} contains a "
                "single label. If this is not intentional, please check.",
                UserWarning,
            )
         .....
        # build dataloader
        dataloader = self._build_dataloader(X, y)
```

File: sktime/classification/deep_learning/rnn/_rnn_torch.py (hunk 3 above)
Comment (Faakhir30) [reply sktime/classification/deep_learning/rnn/_rnn_torch.py:203]: `_build_network` is only called from `_fit` and called after these checks, so yeah, this is fine.

**Comment by Faakhir30** [comment]
Also FYI @RecreationalMath , I havent added `X.shape` checks to old `classification.deep_learning._pytorch.py` considering that we are moving away from that one, and I'm seeing PRs actively migrating modules using it to new `_base_pytorch.py`.

Lmk if we need to add to it too.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 5]
```diff
@@ -601,7 +616,15 @@ def _predict_proba(self, X):
...
+            activation_instance = self.network._instantiate_activation(
+                self.self._validated_activation
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/base/_base_torch.py]: There is a typo here -> `self.self`

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 6]
```diff
@@ -601,7 +616,15 @@ def _predict_proba(self, X):
                 y_pred.append(self.network(**inputs).detach())
         y_pred = cat(y_pred, dim=0)
         # (batch_size, num_outputs)
-        y_pred = Fsoftmax(y_pred, dim=-1)
+
+        if self._validated_activation:
+            activation_instance = self.network._instantiate_activation(
```
Comment (RecreationalMath) [inline sktime/classification/deep_learning/base/_base_torch.py]: There is a bug here.

This is the base classifier class, and `_instantiate_activation` is implemented in the various network classes, and it currently doesn't have a uniform signature as this line assumes. This will break.
Refer the signature below:
- `MCDCNNNetworkTorch._instantiate_activation(self, activation)`
- `MLPNetworkTorch._instantiate_activation(self, layer)`
- `RNNNetworkTorch._instantiate_activation(self)`

I suggest exploring the idea - if activation instantiation can live somewhere else? It is currently present in all the PyTorch network classes and is essentially duplicated code barring a few changes. This is one of the important enhancement ideas (not sure if already mentioned in #8928 or not).

**Review by RecreationalMath — CHANGES_REQUESTED** [review]
I also suggest adding tests for the functionality that is being changed in this PR and being thorough with the testing.
This is an important piece, and it would benefit from having specific tests.

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 6 above)
Comment (Faakhir30) [reply sktime/classification/deep_learning/base/_base_torch.py]: Cool, I'll refactor `instantiate_activation` to something like `networks/utils/pytorch.py` and support activations without breaking existing torch estimators.

Did arrive on this solution on CNN one too. https://github.com/sktime/sktime/pull/9330#discussion_r2886408165

Thanks for pointing out.

**Review by RecreationalMath — CHANGES_REQUESTED** [review]
I suspect the current implementation in this PR would lead to double application of activation, once in `forward` and once in `_predict_proba`. Please look into it.

**Comment by RecreationalMath** [comment]
> Also FYI @RecreationalMath , I havent added `X.shape` checks to old `classification.deep_learning._pytorch.py` considering that we are moving away from that one, and I'm seeing PRs actively migrating modules using it to new `_base_pytorch.py`.
> 
> Lmk if we need to add to it too.

No need. We are moving out of the old interface.

**Comment by RecreationalMath** [comment]
@Faakhir30 I understand that it started as related and small changes, but in future I would suggest keeping the focus of a PR focused on only a single issue. It's easier to manage and review.

**Comment by Faakhir30** [comment]
>I understand that it started as related and small changes, but in future I would suggest keeping the focus of a PR focused on only a single issue. It's easier to manage and review.

Yess, changes of activation have expanded alot, seperated it to #9622 

This one can be reviewed/merged.

**Review by phoeenniixx — APPROVED** [review]
Thanks!
LGTM

**Review by fkiraly — CHANGES_REQUESTED** [review]
I think deduplicating and refactoring the input checks makes a lot of sense - however, the `BaseClassifier` or `BaseRegressor` should already take care of input checks, so adding them again anywhere further down in the inheritance/template extension tree is duplicative. I would simply remove them entirely.

**Comment by Faakhir30** [comment]
> so adding them again anywhere further down in the inheritance/template extension tree is duplicative.

Yes, these checks are indeeed duplicate, removing in next commit.

tested that following warnings do trigger by running locally, quick references:
**X.shape check**
https://github.com/sktime/sktime/blob/6d386a55d67632fd9c8354a4e0c6b1c5923d7581/sktime/classification/deep_learning/base/_base_torch.py#L634-L649

**single lable dataset warning**
https://github.com/sktime/sktime/blob/6d386a55d67632fd9c8354a4e0c6b1c5923d7581/sktime/base/_base_panel.py#L573-L583

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks - when changing scope or content of a PR, please take care to update its description too.
(apart from that, we could merge)

**Review by fkiraly — APPROVED** [review]
